google/cloud/spanner_v1/merge_result_set.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2024 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from concurrent.futures import ThreadPoolExecutor
+from dataclasses import dataclass
+from queue import Queue
+from typing import Any, TYPE_CHECKING
+from threading import Lock, Semaphore
+
+if TYPE_CHECKING:
+    from google.cloud.spanner_v1.database import BatchSnapshot
+
+QUEUE_SIZE_PER_WORKER = 32
+MAX_PARALLELISM = 100
+METADATA_LOCK = Lock()
```

Comment:
This is now a global lock, right? Would it be possible to make it an instance variable for the MergedResultSet class?

Reply:
Any reason why we want it to be an instance of MergedResultSet because it is not used by MergedResultSet class but used just by static `_set_metadata()` method

Reply:
This means that if you have multiple MergedResultSets open, then they will block each other, which is not necessary. So a more correct design would be to have the lock as an instance variable, and the `_set_meta_data` as an instance method. The goal of this lock is to prevent multiple threads from setting/reading the metadata field of a specific `MergedResultSet` at the same time, not to prevent different `MergedResultSet`s from setting their respective metadata fields.

I agree that it is a bit theoretical, as it is unlikely that a user will have a large number of `MergedResultSet`s open at the same time, but putting the lock where it is actually needed will make the code easier to read and understand. Now it seems like it would be a problem if two different `MergedResultSet`s try to set their metadata at the same time.

Reply:
SG, Changed